### ISSUE TITLE
Suggestion:  Perhaps LSP is relevant to Notepad++ ?

### ISSUE THREAD

**Issue by PH0lder** [notepad-plus-plus/notepad-plus-plus#4440]
### Description of the Issue
LSP ( https://langserver.org/ ) is a general way for multiple different test/code editors to support many different development/programming languages.

This seems like it could be something that might be useful for Notepad++ to investigate and eventually include.

If Notepad++ supported LSP, and adding new LSP language servers, then as new languages get invented, and their teams create a LSP server for the new language, they could be "automagically" supported by Notepad++.

**Comment by snow8gaia** [comment]
https://www.reddit.com/r/vim/comments/b3yzq4/a_lsp_client_maintainers_view_of_the_lsp_protocol/
https://old.reddit.com/r/vim/comments/b33lc1/a_guide_to_lsp_auto_completion_in_vim/

**Comment by silvioprog** [comment]
More references:

- NppLSP: experimental plugin for LSP support in Notepad++ - https://github.com/dail8859/NppLSP
- Mature LSP support for Emacs - https://github.com/emacs-lsp/lsp-mode

**Comment by pidgeon777** [comment]
Any chance LSP will be integrated into Notepad++ one day?

I think that even some external plugins like "Python Script" could maybe do the trick...

**Comment by sasumner** [comment]
>Any chance LSP will be integrated into Notepad++ one day?

Well, this issue is "open" :-)

**Comment by ArkadiuszMichalski** [comment]
Next references:
https://microsoft.github.io/language-server-protocol/ - official LSP standard
https://microsoft.github.io/language-server-protocol/implementors/servers/ - list of servers
https://microsoft.github.io/language-server-protocol/implementors/tools/ - list of clients
https://www.toptal.com/javascript/language-server-protocol-tutorial - interesting tutorial
https://www.zeusedit.com/lsp/lsp.html - Language Server Configuration Dialog for Zeus IDE (as example)

This will gives us a lot of possibilities when creating code and it would be good to have it in NPP, at least as a plugin. If someone is tempted to create a plugin, it will certainly gain a large audience.

**Comment by pidgeon777** [comment]
Still hoping one day LSP will be integrated into Notepad++. :)

**Comment by pidgeon777** [comment]
A developer has finally implemented an LSP client for Notepad++ through a plugin:

https://github.com/Ekopalypse/NppLspClient

I encourage everyone to give it a try and report back for any found issue 👍.

**Comment by ColourGrey** [comment]
@pidgeon777: That project suffers from a number of inconvenient issues:

- after a short burst of activity between Oct. 25 and Nov. 21, everything seems to have stopped;
- it is written in V (which is currently an obscure language - I wish it luck, but it still has a long way to go), and it therefore requires both the V compiler and a C compiler to be present.

While the second point is manageable, it together with the first suggest that this is yet another hobby project without any ambition of leaving a mark and becoming mainstream. This is in clear contradiction with NPP, which is a well established, well regarded, mature and serious project. My opinion is backed by the fact that the same author has begun and then abandoned [a similar project, but in Python](https://github.com/Ekopalypse/NppPythonScripts/tree/master/lspclient).

Notice that another another short-lived project with similar goals is [NppLSP](https://github.com/dail8859/NppLSP), by a different author.

This is a major problem with non-paid development: people begin projects enthusiastically, but get bored quickly (or encounter various difficulties that they had overlooked, being overoptimistic) and abandon them.

I believe that if we were to make a poll, adding the LSP client capability to NPP would be the most voted feature, turning NPP into almost an IDE - but one being able to run in less than a dozen megabytes! For the reasons described above, I don't think that we should expect much from good-willed but over-confident volunteers. Everyting seems to depend on how much time @donho would be willing to allocate to this.

**Comment by Ekopalypse** [comment]
Hello @ColourGrey
I'm not sure what to answer other than, no, it hasn't stopped, not yet.
Thank you for your valuable thoughts.

For entertainment a small demo of the plugin.
![lsp_demo](https://user-images.githubusercontent.com/47723516/149093444-afb8e2fa-c355-46ee-8ce1-c6e5f1ebd819.gif)

**Comment by double-standard-you** [comment]
> ### Description of the Issue
> LSP ( https://langserver.org/ ) is a general way for multiple different test/code editors to support many different development/programming languages.
> 
> This seems like it could be something that might be useful for Notepad++ to investigate and eventually include.
> 
> If Notepad++ supported LSP, and adding new LSP language servers, then as new languages get invented, and their teams create a LSP server for the new language, they could be "automagically" supported by Notepad++.

American Shares Truth About Joshua Wong & Hong Kong Democracy
https://www.youtube.com/watch?v=8nXz66Btfl0

American Shares Truth on Xinjiang and Tibet Controversies in China
https://www.youtube.com/watch?v=CldtoYwPtMk

Why is Ukraine the West’s Fault? Featuring John Mearsheimer
https://www.youtube.com/watch?v=JrMiSQAGOS4

Bat coronavirus found in U.S.-funded bio-lab in Ukraine: Russian Defense Ministry
https://www.youtube.com/watch?v=BCydUeHAhzQ

Ukrainians were denied US visas as White House promise help to refugees
https://www.youtube.com/watch?v=BGZhzNFmyhc

Reports of racist treatment against Africans trying to flee Ukraine
https://www.youtube.com/watch?v=JItYOW_uT8s

Fleeing African and Indian students face racism at Ukraine border
https://www.youtube.com/watch?v=ODMOzwI__zs

Canadian RCMP use horses to trample protestors
https://www.youtube.com/watch?v=Usn3CR0E1v0

A bald-faced lie from the Western media corp: Critical “Uyghur Genocide” Questions: A CBC Interview Follow-up
https://www.youtube.com/watch?v=O7IbsPFM2Xg

**Comment by michal-kapala** [comment]
@donho May I ask what the current status of this feature request is? I'm probably one of many LSP-compliant plugin developers coming from editors like VS Code/Neovim/Atom etc. and would love to extend the support for Notepad++ without the need of rewriting the same functionality with a different plugin system.

Has this been reviewed or discussed in any way? It's been here for nearly 5 years now.

**Comment by Kered13** [comment]
I don't believe that @donho is working on this or plans to work on it. However there are two projects working on developing LSP plugins for NPP, one by @Ekopalypse and one by myself, both linked above. I can't speak for @Ekopalypse, but my project is still on off-and-on development. Any contributions would certainly be appreciated.

**Comment by Ekopalypse** [comment]
I second what @Kered13 shared.
